src/lib/panorama.ts

```diff
@@ -259,147 +267,187 @@ function getTileUrl(pano: Panorama, x: number, y: number, zoom: number) {
-			tile.renderOrder = 1;
-			panorama.add(tile);
-		}
-	}
+		// The image is drawn twice here in order to accomodate the yaw shift.
+		// Once from the shifted point to the end of the canvas...
+		envContext.drawImage(image, shift, 0, cropWidth - shift, cropHeight, 0, 0, cropWidth - shift, cropHeight);
+		// and again from the start of the canvas to the shift point.
+		envContext.drawImage(image, 0, 0, shift, cropHeight, cropWidth - shift, 0, shift, cropHeight);
+		// This approach isn't perfect, the environment isn't going to correct for pitch and roll.
+		// But it doesn't have to be, the environment is only used for lighting and reflections.
+
+		envTexture.needsUpdate = true;
+
+		const envMap = pmremGenerator.fromEquirectangular(envTexture).texture;
+		scene.environment = envMap;
+	};
 }
 
-function setSceneEnvironment(
-	pano: Panorama,
+const TAU = 2 * Math.PI;
+const geometryCache = {};
+
+type PanoMetadata = NonNullable<Awaited<ReturnType<typeof getPanoMetadataFromId>>>;
+export async function renderPanoFromMetadata(
+	meta: PanoMetadata,
 	scene: THREE.Scene,
 	pmremGenerator: THREE.PMREMGenerator,
-	width: number,
-	height: number,
-) {
-	new Promise((resolve, reject) => {
-		const img = new Image();
-		img.crossOrigin = 'anonymous';
-		img.onload = () => resolve(img);
-		img.onerror = reject;
-		img.src = getTileUrl(pano, 0, 0, 0);
-	}).then((image: HTMLImageElement) => {
-		const canvas = document.createElement('canvas');
-		canvas.width = width;
-		canvas.height = height;
-		const ctx = canvas.getContext('2d')!;
-		const shift = Math.floor((90 / 360) * width);
-		ctx.drawImage(image, shift, 0, width - shift, height, 0, 0, width - shift, height);
-		ctx.drawImage(image, 0, 0, shift, height, width - shift, 0, shift, height);
-
-		const envTexture = new THREE.CanvasTexture(canvas);
-		envTexture.mapping = THREE.EquirectangularReflectionMapping;
-		envTexture.colorSpace = THREE.SRGBColorSpace;
-		const envMap = pmremGenerator.fromEquirectangular(envTexture).texture;
-		scene.environment = envMap;
+	heading: number,
+	maxZoom: number,
+): Promise<THREE.Group<THREE.Object3DEventMap>> {
+	return new Promise((resolve, reject) => {
+		const zoom = Math.min(meta.maxZoom, maxZoom);
+
+		const sliceHorizontal = TAU / (meta.zoomLevels[zoom].cropWidth / meta.tileWidth);
+		const sliceVertical = Math.PI / (meta.zoomLevels[zoom].cropHeight / meta.tileHeight);
+
+		const panorama = new THREE.Group();
+		panorama.name = 'panorama';
+		panorama.rotation.order = 'YZX';
+		panorama.rotation.set(
+			THREE.MathUtils.degToRad((360 + meta.roll) % 360),
+			THREE.MathUtils.degToRad((360 + (90 - meta.heading) + heading) % 360),
+			THREE.MathUtils.degToRad((360 + (meta.tilt - 90)) % 360),
+		);
+
+		const loadingManager = new THREE.LoadingManager();
+		loadingManager.onLoad = () => {
+			newPano = panorama;
+			scene.add(panorama);
+			panoFadeTime = performance.now();
+			setSceneEnvironment(meta, scene, pmremGenerator, heading);
```

Comment:
Is it required we wait for the tile here to load to try and load the scene environment too?

I guess you were trying to only load the scene environment if you knew the tiles would load correctly. But I'd say we can parallelize this by calling `setSceneEnvironment()` alongside `renderPanoFromMetadata()` and just quietly failing in `setSceneEnvironment()`.